Add infinite reconnection attempts.

sphh opened the pull request:
If the parameter `reconnect_attempts` in `MopidyClient.__init__()` is set to `None`, try to connect indefinitely.

sphh committed:
Add infinite reconnection attempts.

If the parameter `reconnect_attempts` is set to `None`, try to connect indefinitely.

sphh commented:
@SvineruS: May I ask what your intention is? Are you considering dealing with the PRs? I use this library in a program, which implements the MPRIS2 DBus protocol for a Mopidy instance, which is running remotely. I want to publish it, but would need at least the functionality of this PR. I am not too happy to make new fork and strongly believe, that there should be one, just one, library dealing with one task. Thanks.

SvineruS commented:
Hello. Sorry, I haven't seen PR alerts. I will look at them when there 
is time (probably tomorrow). If you have a telegram, write to me there, 
otherwise I may forget) <https://t.me/svinerus>

В Пт, апр 23, 2021 at 11:07, sphh ***@***.***> 
написал:
> @SvineruS <https://github.com/SvineruS>: May I ask what your 
> intention is? Are you considering dealing with the PRs? I use this 
> library in a program, which implements the MPRIS2 DBus protocol for a 
> Mopidy instance, which is running remotely. I want to publish it, but 
> would need at least the functionality of this PR. I am not too happy 
> to make new fork and strongly believe, that there should be one, just 
> one, library dealing with one task. Thanks.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/SvineruS/mopidy-async-client/pull/1#issuecomment-825829830>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AI5DAPWQXG66MRQOSFXQ2DLTKGZOVANCNFSM42BQUXUQ>.
>

sphh commented:
Thanks for looking into them.

Strange, that you have not received any PR alerts …

Sorry, no telegram, I use matrix.

SvineruS commented:
> If the parameter `reconnect_attempts` in `MopidyClient.__init__()` is set to `None`, try to connect indefinitely.

I like the goal. but I hate while loops. how about ```self._reconnect_attempts = reconnect_attempts or sys.maxsize```?

sphh commented:
I thought about this, because then I could have kept the for loop, resulting in fewer changes. But I don't believe, that setting the maximum number of loops for infinite connection attempts is a sound approach, even if it is the sysmax. Firstly, this number is system dependent, secondly there is the (theoratical) case that this maximum is reached, and thirdly it is not the maximum size (`sys.maxsize+sys.maxsize` returns a valid number). BTW that's also the reason, why I packed the incrementation of `i` into an `if` clause.
Actually I believe, that's exactly the reason, why `while` loops were invented …

SvineruS commented:
> secondly there is the (theoratical) case that this maximum is reached  

I am not very good at designing such systems, but, in my opinion, an infinite loop of queries is bad. I believe that the number of attempts should be limited (like in the ping utility). 

by the way, for an infinite loop, you still have to change the message in the log.

sphh committed:
Add different log message for infinite reconnection attempts.

Also move the logging statement out of the `try: except:` statement, because that should only catch errors of `self.connect()` and not of the logging system.

sphh commented:
You are right. But this now conflicts with #2, so I also changed the second `logging.info()` statement to `logger.info()` statement. #2 is now partly obsolete.

I also pulled the logging statement out of the `try: except:` clause, because that clause should only catch exception of the connect statement and not of the logging system.

sphh commented:
Ah, to reply to your "in infinite loop of queries is bad": Why do you think so? Isn't that the reason, why `while` loops exist? I would have thought, that Guido would not have implemented them, if they are bad and I have seen them in many, many programs. But I am happy to see "the other side"!

SvineruS commented:
now I would like the independent side to resolve our conflict, but you are probably the only one who uses this lib)

the problem is not with while (although I still hate it), but that in my opinion an infinite number of attempts is bad. I think it's worth looking at other products and their solutions.

sphh commented:
Ah, I see. I always thought, it's because of the `while` loop. Because in that case, I would have suggested to use something like:
```python3
while True:
    try:
        self.connect()
    except OSError:
        …
    if i < self._reconnect_attempt:
        break
```
Still a `while` loop, but possibly cleaner.

As why I need infinite connects: My program connects the MPRIS2 DBus to a Mopidy instance running remotely (let's call this program `mopidy-MPRIS2` for the moment). For example the Sound Applet is using the MPRIS2 framework to control media players. Most of them are running on the desktop, but there is one ([mpDris2](https://github.com/eonpatapon/mpDris2)), which does exactly the same as my program, but with `mpd`. Now you could argue, that you have to start the `mopidy-MPRIS2` program whenever you want to control the remote mopidy. Then the program should disappear, if the remote mopidy disappears. But since the remote mopidy is a long-running server process, it makes much sense, that `mopidy-MPRIS2` tries to connect to this server forever. It would be a bit weird, if `mopidy-MPRIS2` disappears, whenever the connection disappears (thinking flaky WiFi connections). This would be a similar to an email client, which closes the program, if it cannot reach the mail server for 3 times in a row …

sphh commented:
BTW who would be a independent side? The developers of `mopidy`? We can certainly ask them. Should I do it?

SvineruS commented:
okay. but did you know that a loop with sys.maxsize will last for millions of years? :)
this solution is much more elegant than while, please, let's use it :))

SvineruS commented:
> BTW who would be a independent side? The developers of `mopidy`? We can certainly ask them. Should I do it?

no, just a random third developer. if the topic is still not closed.

sphh commented:
1. That's what they said last century, when they thought, that they don't have to use the century in front of the year, resulting in a frenzy before the new year's day of 2000 (https://en.wikipedia.org/wiki/Year_2000_problem)
2. I still don't know, what's so bad about a `while` loop. Please do convince me!
3. Is it now again about the `while` loop or about the infinite connection attempts? I get confused.

SvineruS commented:
1. I don't think the program will work for a million years)
2. it's ugly)
3. now it's about while again, since maxsize is also nearly infinite

SvineruS commented:
and you always can do an outer loop if you really want, lol

sphh commented:
1. See the 2yk bug.
2. See my proposal to use `while: break`. Is that better? Or is it just a general aversion against `while` loops?
3. But exactly that's the point: It is not infinite, just nearly. "Just do it properly," I would say.
4. Yes, I could use an outer loop. But that's really ugly.

If you really don't want to have a while loop in your program, that's ok. Than I will fork it and increase the number of very similar Python libraries by another one. I find that sad. (I could never understand, why there are so many very similar libraries around, but now I might know why.)

Just tell me your decision.

SvineruS commented:
shit, I share your opinion about the fork and the same libraries. if you so want this while I will make concessions. but he's ugly.

sphh commented:
Do you like the `while: break` construct from https://github.com/SvineruS/mopidy-async-client/pull/1#issuecomment-825915653 a tiny bit better?

SvineruS commented:
I do not know. just the cleaner the code the better. I'm too lazy to decide, I'm high.

sphh committed:
Update client.py

sphh committed:
Update client.py

sphh commented:
@SvineruS: Are you better today?

I thought about it and came up with the following solution, which I believe is clean and easy to understand:
```python3
    async def _reconnect(self):

        async def _reconnect_():
            # Try to connect for a maximum of '_reconnect_attempts'
            await self.disconnect()
            for i in range(self._reconnect_attempts - 1):
                logger.info("try to reconnect. attempt %s / %s", i, self._reconnect_attempts)
                try:
                    await self.connect()
                except OSError:
                    logger.info("reconnect failed. new attempt in %s sec", self._reconnect_timeout)
                    await asyncio.sleep(self._reconnect_timeout)
                else:
                    return
            await self.connect()  # not catching last attempt

        async def _reconnect_forever_():
            # Try to connect forever
            await self.disconnect()
            while True:
                logger.info("try to reconnect.")
                try:
                    await self.connect()
                except OSError:
                    logger.info("reconnect failed. new attempt in %s sec", self._reconnect_timeout)
                    await asyncio.sleep(self._reconnect_timeout)
                else:
                    return

        if self._reconnect_attempts is None:
            self._loop.create_task(_reconnect_forever_())
        else:
            self._loop.create_task(_reconnect_())  # this task will be closed so creating new one
```
What do you think?

SvineruS commented:
well fuck, this is even more code for one action ¯\_(ツ)_/¯

sphh commented:
Yes, but much, much cleaner! And very easy to read and everyone knows at a first glance, what is going on. The challenge should not be to write the fewest lines of codes, but the easiest to read and understand.

sphh commented:
Sorry, I just saw, that you went down the `sys.maxsize` route. You just introduced three bugs:
1. `sys` is not imported.
2. `range(sys.maxsize)` is definitely not an infinite loop.

Do I really have to fork this??

SvineruS commented:
1. fix
2. i dont think so
3. where?
4. idk

sphh commented:
1. Just to import a new module, because one hates `while` loops is a bit weird.
2. Every loop, which has an upper limit (however large) is per definition not an infinite loop.
3. The usual channels.
4. What do you not know?

SvineruS commented:
3. ?
4. >Do I really have to fork this??

sphh commented:
I don't know, if you have to fork it. But it looks like that there is no way for me do _not_ fork it.
